### PR TITLE
Feature/default configurable value

### LINKED ISSUE #338: Show available values for configurable rule parameters
The only way to get information what values are accepted for a configurable now it to provide incorrect value and get an exception with such information.
I would like to have available values listed when calling `-lc / --list-configurables` option and also have a default value marked.

### PR DESCRIPTION
**Pull request by bhirsz**
Closes #338

### CODE REVIEW

File: tests/utest/test_listing_rules.py [hunk 1]
```diff
@@ -135,15 +136,15 @@ def test_list_filtered(self, robocop_pre_load, msg_0101, msg_0102_0204, capsys):
...
         robocop_pre_load.config.list_configurables = robocop.config.translate_pattern('*')
-        init_empty_checker(robocop_pre_load, msg_0101_config)
+        init_empty_checker(robocop_pre_load, msg_0101_config, conf_param=1001)
         with pytest.raises(SystemExit):
             robocop_pre_load.list_checkers()
         out, _ = capsys.readouterr()
         assert out == "All rules have configurable parameter 'severity'. " \
                       "Allowed values are:\n    E / error\n    W / warning\n    I / info\n" \
                       "Rule - 0101 [W]: some-message: Some description (enabled)\n" \
                       "    Available configurable(s) for this rule:\n" \
```
Comment (mnojek) [inline tests/utest/test_listing_rules.py:146]: What about removing `Available configurable(s) for this rule:` ?

Comment (bhirsz) [reply tests/utest/test_listing_rules.py:146]: I can remove it - I will update PR

Comment (mnojek) [reply tests/utest/test_listing_rules.py]: Then you can also decrease the indentation by 1 "tab"

Comment (bhirsz) [reply tests/utest/test_listing_rules.py]: Removed, decreased indentation - but I've left 'Available configurable(s)' for exception that is throwed when you're trying to configure rule with non existing parameter.

File: robocop/run.py [hunk 2]
```diff
@@ -271,7 +270,8 @@ def configure_checkers_or_reports(self):
                     if configurable is None:
                         available_conf = msg.available_configurables()
                         raise robocop.exceptions.ConfigGeneralError(
-                            f"Provided param '{param}' for rule '{rule_or_report}' does not exist. {available_conf}")
+                            f"Provided param '{param}' for rule '{rule_or_report}' does not exist. "
+                            f"Available configurable(s) for this rule:\n    {available_conf}")
```
Comment (mnojek) [inline robocop/run.py:274]: What about:
```suggestion
                            f"Available configurable(s) for this rule:\n"
                            f"    {available_conf}"
                            )
```
?

File: robocop/run.py [hunk 3]
```diff
@@ -174,20 +174,19 @@ def list_checkers(self):
...
             else:
-                if not rule_by_id[rule_id].matches_pattern(self.config.list_configurables):
+                if not rule_def.matches_pattern(self.config.list_configurables):
                     continue
-                configurables = rule_by_id[rule_id].available_configurables(include_severity=False)
-                configurables = f'\n    {configurables}' if configurables else ''
+                configurables = rule_def.available_configurables(include_severity=False, checker=checker)
                 if configurables:
-                    print(f"{rule_by_id[rule_id]}{configurables}")
+                    print(f"{rule_def}\n    {configurables}")
```
Comment (mnojek) [inline robocop/run.py:189]: I think for better visual understanding of what is being printed, it's better to call it like this:
```suggestion
                    print(f"{rule_def}\n"
                          f"    {configurables}")
```

File: robocop/checkers/lengths.py [hunk 4]
```diff
@@ -196,7 +196,7 @@ class NumberOfReturnedArgsChecker(VisitorChecker):
             "number-of-returned-values",
             "Too many return values (%d/%d)",
             RuleSeverity.WARNING,
-            ("max_returns", "max_returns", int)
+            ("max_returns", "max_returns", int, 'allowed number of returned values from keyword')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:199]: Split into multiple lines

File: robocop/checkers/misc.py [hunk 5]
```diff
@@ -151,13 +151,25 @@ class ConsistentAssignmentSignChecker(VisitorChecker):
             "inconsistent-assignment",
             "The assignment sign is not consistent through the file. Expected '%s' but got '%s' instead",
             RuleSeverity.WARNING,
-            ('assignment_sign_type', 'keyword_assignment_sign_type', parse_assignment_sign_type)
+            (
+                'assignment_sign_type',
+                'keyword_assignment_sign_type',
+                parse_assignment_sign_type,
+                "value can be one of: 'autodetect' (default), 'equal_sign' ('='), 'none' ('') "
```
Comment (mnojek) [inline robocop/checkers/misc.py:158]: What about "possible values" or "accepted values"?

File: robocop/checkers/misc.py [hunk 6]
```diff
@@ -151,13 +151,25 @@ class ConsistentAssignmentSignChecker(VisitorChecker):
...
+                "or space_and_equal_sign (' =') "
```
Comment (mnojek) [inline robocop/checkers/misc.py:159]: Remove trailing whitespace inside str

File: robocop/checkers/lengths.py (hunk 4 above)
Comment (mnojek) [reply robocop/checkers/lengths.py:199]: Also what about other rules? 😃

Comment (bhirsz) [reply robocop/checkers/lengths.py:199]: It was just for testing :D I can go through all rules with configurables and update the meta info (its optional BTW)

Comment (mnojek) [reply robocop/checkers/lengths.py:199]: I would add it in the places where it's not obvious. Or maybe everywhere? Not sure...

Comment (mnojek) [reply robocop/checkers/lengths.py:199]: Or only where there is enumerated value

**Comment by bhirsz** [comment]
Closes #351 

I've also updated our auto generated docs - you can see them at https://robocop.readthedocs.io/en/feature-default_configurable_value/index.html

File: robocop/checkers/lengths.py [hunk 7]
```diff
@@ -148,7 +148,7 @@ class LineLengthChecker(RawFileChecker):
             "line-too-long",
             "Line is too long (%d/%d)",
             RuleSeverity.WARNING,
-            ("line_length", "max_line_length", int)
+            ("line_length", "max_line_length", int, 'maximum allowed number of chars in one line')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:151]:
```suggestion
            ("line_length", "max_line_length", int, 'number of characters allowed in one line')
```

File: robocop/checkers/lengths.py [hunk 8]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('max_lines', 'file_max_lines', int)
+            ('max_lines', 'file_max_lines', int, 'maximum allowed number of lines')
         ),
         "0507": (
             "too-many-arguments",
             "Keyword has too many arguments (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_args', 'keyword_max_args', int)
+            ('max_args', 'keyword_max_args', int, 'maximum allowed number of arguments')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:59]:
```suggestion
            ('max_args', 'keyword_max_args', int, 'number of arguments a keyword can take')
```

File: robocop/checkers/lengths.py [hunk 9]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('max_calls', 'testcase_max_calls', int)
+            ('max_calls', 'testcase_max_calls', int, 'maximum allowed number of keywords inside')
         ),
         "0506": (
             "file-too-long",
             "File has too many lines (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_lines', 'file_max_lines', int)
+            ('max_lines', 'file_max_lines', int, 'maximum allowed number of lines')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:53]:
```suggestion
            ('max_lines', 'file_max_lines', int, 'number of lines allowed in a file')
```

File: robocop/checkers/lengths.py [hunk 10]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('max_len', 'testcase_max_len', int)
+            ('max_len', 'testcase_max_len', int, 'maximum allowed number of lines')
         ),
         "0505": (
             "too-many-calls-in-test-case",
             "Test case have too many keywords inside (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_calls', 'testcase_max_calls', int)
+            ('max_calls', 'testcase_max_calls', int, 'maximum allowed number of keywords inside')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:47]:
```suggestion
            ('max_calls', 'testcase_max_calls', int, 'number of keyword calls allowed in a test case')
```

File: robocop/checkers/lengths.py [hunk 11]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('max_calls', 'keyword_max_calls', int)
+            ('max_calls', 'keyword_max_calls', int, 'maximum allowed number of keywords inside')
         ),
         "0504": (
             "too-long-test-case",
             "Test case is too long (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_len', 'testcase_max_len', int)
+            ('max_len', 'testcase_max_len', int, 'maximum allowed number of lines')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:41]:
```suggestion
            ('max_len', 'testcase_max_len', int, 'number of lines allowed in a test case')
```

File: robocop/checkers/lengths.py [hunk 12]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('min_calls', 'keyword_min_calls', int)
+            ('min_calls', 'keyword_min_calls', int, 'minimum required number of keywords inside')
         ),
         "0503": (
             "too-many-calls-in-keyword",
             "Keyword have too many keywords inside (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_calls', 'keyword_max_calls', int)
+            ('max_calls', 'keyword_max_calls', int, 'maximum allowed number of keywords inside')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:35]:
```suggestion
            ('max_calls', 'keyword_max_calls', int, 'number of keyword calls allowed in a keyword')
```

File: robocop/checkers/lengths.py [hunk 13]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
...
             RuleSeverity.WARNING,
-            ('max_len', 'keyword_max_len', int)
+            ('max_len', 'keyword_max_len', int, 'maximum allowed number of lines')
         ),
         "0502": (
             "too-few-calls-in-keyword",
             "Keyword have too few keywords inside (%d/%d)",
             RuleSeverity.WARNING,
-            ('min_calls', 'keyword_min_calls', int)
+            ('min_calls', 'keyword_min_calls', int, 'minimum required number of keywords inside')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:29]:
```suggestion
            ('min_calls', 'keyword_min_calls', int, 'minimum number of keyword calls required in a keyword')
```

File: robocop/checkers/lengths.py [hunk 14]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
             "too-long-keyword",
             "Keyword is too long (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_len', 'keyword_max_len', int)
+            ('max_len', 'keyword_max_len', int, 'maximum allowed number of lines')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:23]:
```suggestion
            ('max_len', 'keyword_max_len', int, 'number of lines allowed in a keyword')
```

File: robocop/checkers/lengths.py [hunk 15]
```diff
@@ -203,7 +203,7 @@ class NumberOfReturnedArgsChecker(VisitorChecker):
             "number-of-returned-values",
             "Too many return values (%d/%d)",
             RuleSeverity.WARNING,
-            ("max_returns", "max_returns", int)
+            ("max_returns", "max_returns", int, 'allowed number of returned values from keyword')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:206]:
```suggestion
            ("max_returns", "max_returns", int, 'allowed number of returned values from a keyword')
```

File: robocop/checkers/misc.py [hunk 16]
```diff
@@ -151,31 +151,43 @@ class ConsistentAssignmentSignChecker(VisitorChecker):
             "inconsistent-assignment",
             "The assignment sign is not consistent through the file. Expected '%s' but got '%s' instead",
             RuleSeverity.WARNING,
-            ('assignment_sign_type', 'keyword_assignment_sign_type', parse_assignment_sign_type)
+            (
+                'assignment_sign_type',
+                'keyword_assignment_sign_type',
+                parse_assignment_sign_type,
+                "possible values: 'autodetect' (default), 'equal_sign' ('='), 'none' ('') "
+                "or space_and_equal_sign (' =')"
```
Comment (mnojek) [inline robocop/checkers/misc.py:159]:
```suggestion
                "possible values: 'autodetect' (default), 'none' (''), 'equal_sign' ('=') "
                "or space_and_equal_sign (' =')"
```

File: robocop/checkers/misc.py [hunk 17]
```diff
@@ -151,31 +151,43 @@ class ConsistentAssignmentSignChecker(VisitorChecker):
...
             "inconsistent-assignment-in-variables",
             "The assignment sign is not consistent inside the variables section. Expected '%s' but got '%s' instead",
             RuleSeverity.WARNING,
-            ('assignment_sign_type', 'variables_assignment_sign_type', parse_assignment_sign_type)
+            (
+                'assignment_sign_type',
+                'variables_assignment_sign_type',
+                parse_assignment_sign_type,
+                "possible values: 'autodetect' (default), 'equal_sign' ('='), 'none' ('') "
+                "or space_and_equal_sign (' =')"
```
Comment (mnojek) [inline robocop/checkers/misc.py:171]:
```suggestion
                "possible values: 'autodetect' (default), 'none' (''), 'equal_sign' ('=') "
                "or space_and_equal_sign (' =')"
```

File: robocop/checkers/misc.py [hunk 18]
```diff
@@ -151,31 +151,43 @@ class ConsistentAssignmentSignChecker(VisitorChecker):
...
+        self.variables_assignment_sign_type = 'autodetect'
         self.keyword_expected_sign_type = None
         self.variables_expected_sign_type = None
         super().__init__()
 
     def visit_File(self, node):  # noqa
         self.keyword_expected_sign_type = self.keyword_assignment_sign_type
         self.variables_expected_sign_type = self.variables_assignment_sign_type
-        if self.keyword_assignment_sign_type is None or self.variables_assignment_sign_type is None:
+        if self.keyword_assignment_sign_type == 'autodetect' or self.variables_assignment_sign_type == 'autodetect':
```
Comment (mnojek) [inline robocop/checkers/misc.py:186]:
```suggestion
        if 'autodetect' in [self.keyword_assignment_sign_type, self.variables_assignment_sign_type]:
```

File: robocop/checkers/naming.py [hunk 19]
```diff
@@ -23,7 +23,7 @@ class InvalidCharactersInNameChecker(VisitorChecker):
             "invalid-char-in-name",
             "Invalid character %s in %s name",
             RuleSeverity.WARNING,
-            ('invalid_chars', 'invalid_chars', set)
+            ('invalid_chars', 'invalid_chars', set, 'set of chars not allowed in name')
```
Comment (mnojek) [inline robocop/checkers/naming.py:26]:
```suggestion
            ('invalid_chars', 'invalid_chars', set, 'set of characters not allowed in a name')
```

File: robocop/checkers/spacing.py [hunk 20]
```diff
@@ -72,25 +72,25 @@ class EmptyLinesChecker(VisitorChecker):
             "empty-lines-between-sections",
             "Invalid number of empty lines between sections (%d/%d)",
             RuleSeverity.WARNING,
-            ("empty_lines", "empty_lines_between_sections", int)
+            ("empty_lines", "empty_lines_between_sections", int, 'required number of empty lines')
```
Comment (mnojek) [inline robocop/checkers/spacing.py:75]:
```suggestion
            (
                "empty_lines",
                "empty_lines_between_sections",
                int,
                "number of empty lines required between sections"
            )
```

File: robocop/checkers/spacing.py [hunk 21]
```diff
@@ -72,25 +72,25 @@ class EmptyLinesChecker(VisitorChecker):
...
         ),
         "1004": (
             "empty-lines-between-test-cases",
             "Invalid number of empty lines between test cases (%d/%d)",
             RuleSeverity.WARNING,
-            ("empty_lines", "empty_lines_between_test_cases", int)
+            ("empty_lines", "empty_lines_between_test_cases", int, 'required number of empty lines')
```
Comment (mnojek) [inline robocop/checkers/spacing.py:81]:
```suggestion
            (
                "empty_lines",
                "empty_lines_between_test_cases",
                int,
                "number of empty lines required between test cases"
            )
```

File: robocop/checkers/spacing.py [hunk 22]
```diff
@@ -72,25 +72,25 @@ class EmptyLinesChecker(VisitorChecker):
...
         ),
         "1005": (
             "empty-lines-between-keywords",
             "Invalid number of empty lines between keywords (%d/%d)",
             RuleSeverity.WARNING,
-            ("empty_lines", "empty_lines_between_keywords", int)
+            ("empty_lines", "empty_lines_between_keywords", int, 'required number of empty lines')
```
Comment (mnojek) [inline robocop/checkers/spacing.py:87]:
```suggestion
            (
                "empty_lines",
                "empty_lines_between_keywords",
                int,
                "number of empty lines required between keywords"
            )
```

File: robocop/checkers/spacing.py [hunk 23]
```diff
@@ -72,25 +72,25 @@ class EmptyLinesChecker(VisitorChecker):
...
         ),
         "1009": (
             "empty-line-after-section",
             "Too many empty lines after section header (%d/%d)",
             RuleSeverity.WARNING,
-            ("empty_lines", "empty_lines_after_section_header", int)
+            ("empty_lines", "empty_lines_after_section_header", int, 'allowed number of empty lines')
```
Comment (mnojek) [inline robocop/checkers/spacing.py:93]:
```suggestion
            (
                "empty_lines",
                "empty_lines_after_section_header",
                int,
                "number of empty lines allowed after section header"
            )
```

File: robocop/checkers/lengths.py [hunk 24]
```diff
@@ -20,43 +20,43 @@ class LengthChecker(VisitorChecker):
             "too-long-keyword",
             "Keyword is too long (%d/%d)",
             RuleSeverity.WARNING,
-            ('max_len', 'keyword_max_len', int)
+            ('max_len', 'keyword_max_len', int, 'number of lines allowed in a keyword')
```
Comment (mnojek) [inline robocop/checkers/lengths.py:23]: What about writing all configurables like this?
```suggestion
            (
                'max_len',
                'keyword_max_len',
                int,
                'number of lines allowed in a keyword'
            )
```

**Review by mnojek — APPROVED** [review]

**Comment by bhirsz** [comment]
Final summary:

* list default value of configurable

* remove 'Available configurable(s)' message from --list-configurables

* refactor listing configurables

* update docs date

* refactor docs

* change ConsistentAssignmentSignChecker default param values to 'autodetect' from None

* add metavars to configurables

* typos have -> has

* split configurables to multi lines

**Pull request merged**